forum/search/typesense.py

```diff
@@ -0,0 +1,377 @@
+"""
+Typesense backend for searching comments and threads.
+"""
+
+from typing import Any, Optional
+
+from bs4 import BeautifulSoup
+from django.conf import settings
+from django.core.paginator import Paginator
+from typesense import Client
+from typesense.types.collection import CollectionCreateSchema
+from typesense.types.document import DocumentSchema, SearchParameters
+from typesense.exceptions import ObjectNotFound
+
+from forum.backends.mysql.models import Comment, CommentThread
+from forum.constants import FORUM_MAX_DEEP_SEARCH_COMMENT_COUNT
+from forum.search.base import (
+    BaseDocumentSearchBackend,
+    BaseIndexSearchBackend,
+    BaseSearchBackend,
+    BaseThreadSearchBackend,
+)
+
+_TYPESENSE_CLIENT: Client | None = None
+
+
+def get_typesense_client() -> Client:
+    """
+    Return a singleton Typesense client instance.
+    """
+    global _TYPESENSE_CLIENT
+    if _TYPESENSE_CLIENT is None:
+        _TYPESENSE_CLIENT = Client(
+            {
+                "api_key": settings.TYPESENSE_API_KEY,
+                "nodes": [settings.TYPESENSE_URL],
```

Comment:
Should we do something like this? https://github.com/openedx/edx-search/blob/fcfda1cc830ed37b04d30e431406277a4c833e6b/search/typesense.py#L39

Reply:
Agreed, let's change it to `TYPESENSE_URLS`. :+1: I can open a PR to tutor-contrib-typesense that makes that change.

Reply:
@bradenmacdonald https://github.com/open-craft/tutor-contrib-typesense/pull/3 is ready for review :)